### PR TITLE
4. Applied Futurize stage 2 across the entire repo.

### PR DESCRIPTION
**Pull request by amakhaba**
This PR does the following ([Apply Futurize stage 2](https://github.com/ska-sa/katdeveloper/wiki/Investigate-a-way-to-improve-Py3-migration-feedback-loop-on-tox#futurize-stage-2)):

- Applies Futurize stage 2 across the entire repository.
- Update document generation in the `Jenkinsfile`.
- Added Py3 support on `setup.py`.

*Definition of Done Checklist*

- [ ] Unit tested (coded, passed, included)?
- [x] Requested at least 2 reviewers?

JIRA: [MT-738](https://skaafrica.atlassian.net/browse/MT-738)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -65,11 +65,11 @@ pipeline {
 
                 stage ('py36') {
                     steps {
-                        echo "Not yet implemented."
-                        // echo "Running nosetests on Python 3.6"
-                        // sh 'python3.6 -m pip install . -U --user'
-                        // sh 'python3.6 -m pip install nose_xunitmp --user'
-                        // sh "python3.6 setup.py nosetests --with-xunitmp --with-xcoverage --cover-package=${KATPACKAGE}"
+                        //echo "Not yet implemented."
```
Comment (mmphego) [inline Jenkinsfile]: Since Py3 testing is enabled, I do not think this is still relevant.

File: Jenkinsfile [hunk 2]
```diff
@@ -65,11 +65,11 @@ pipeline {
...
+                         echo "Running nosetests on Python 3.6"
+                         sh 'python3.6 -m pip install . -U --user'
```
Comment (mmphego) [inline Jenkinsfile]: Please ensure that Py3 and pip3 are installed on the container before envoking them.
As I ain't sure if pip3 is installed.

File: tango_simlib/utilities/base_parser.py [hunk 3]
```diff
@@ -2,17 +2,21 @@
 from __future__ import division
 from __future__ import absolute_import
 
+
```
Comment (mmphego) [inline tango_simlib/utilities/base_parser.py]: Please move above imports below the copyright information.

File: tango_simlib/utilities/base_parser.py [hunk 4]
```diff
@@ -2,17 +2,21 @@
...
 #                                                                                       #
 # BSD license - see LICENSE.txt for details                                             #
 #########################################################################################
-import abc
+from future import standard_library
+standard_library.install_aliases()
 
+from builtins import object
+import abc
+from future.utils import with_metaclass
```
Comment (mmphego) [inline tango_simlib/utilities/base_parser.py]: Please reorder the imports as per [CAM - Python style guide](https://docs.google.com/document/d/1aZoIyR9tz5rCWr2qJKuMTmKp2IzHlFjrCFrpDDHFypM/edit#heading=h.y0sdkqkmrbb3)

File: tango_simlib/utilities/fandango_json_parser.py [hunk 5]
```diff
@@ -45,7 +49,7 @@ def parse(self, json_file):
         with open(json_file) as dev_data_file:
             device_data = json_load_byteified(dev_data_file)
 
-        for data_component, elements in device_data.items():
+        for data_component, elements in list(device_data.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/fandango_json_parser.py]: No need to list the dict items.

```suggestion
        for data_component, elements in device_data.items():
```

File: tango_simlib/utilities/fandango_json_parser.py [hunk 6]
```diff
@@ -63,10 +67,10 @@ def preprocess_command_types(self, command_data):
         types and rename the command properties to match with the keyword arguments of
         the command signature.
         """
-        for cmd_name, cmd_config in command_data.items():
+        for cmd_name, cmd_config in list(command_data.items()):
             self._device_commands[cmd_name] = {}
 
-            for cmd_prop, cmd_prop_value in cmd_config.items():
+            for cmd_prop, cmd_prop_value in list(cmd_config.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/fandango_json_parser.py]: Same here, see comment above.

File: tango_simlib/utilities/fandango_json_parser.py [hunk 7]
```diff
@@ -134,7 +138,7 @@ def update_property_data(self, property_data):
                     ]
                 ),
             )
-            for prop, prop_val in property_data.items()
+            for prop, prop_val in list(property_data.items())
```
Comment (mmphego) [inline tango_simlib/utilities/fandango_json_parser.py]: See comment above regarding the use of `list`.

File: tango_simlib/utilities/helper_module.py [hunk 8]
```diff
@@ -2,11 +2,15 @@
 from __future__ import division
 from __future__ import absolute_import
 
+
```
Comment (mmphego) [inline tango_simlib/utilities/helper_module.py]: Place above imports below license info.

File: tango_simlib/utilities/helper_module.py [hunk 9]
```diff
@@ -119,7 +123,7 @@ def append_device_to_db_file(
     db = Database(db_file_name)
     # Patched the property dict to avoid a PyTango bug
     patched = dict(
-        (key, value if value != "" else " ") for key, value in properties.items()
+        (key, value if value != "" else " ") for key, value in list(properties.items())
```
Comment (mmphego) [inline tango_simlib/utilities/helper_module.py]: This can be rewritten as suggested. 


## TL;DR
>With CPython 2.7, using dict() to create dictionaries takes up to
6 times longer and involves more memory allocation operations than the
literal syntax. Use {} to create dictionaries, especially if you
are pre-populating them, unless the literal syntax does not work for
your case.

Ref: [The Performance Impact of Using dict() Instead of {} in CPython 2.7](https://doughellmann.com/blog/2012/11/12/the-performance-impact-of-using-dict-instead-of-in-cpython-2-7-2/)

In Python you should always use literal syntax whenever possible, this improves performance and efficiency - See image below.

image

```suggestion
    patched = {
        key: value if value != "" else " " for key, value in properties.items()
    }
```

File: tango_simlib/utilities/sim_sdd_xml_parser.py [hunk 10]
```diff
@@ -351,12 +355,12 @@ def _convert_attribute_info(self):
...
         monitoring_pts = {}
-        for mpt_name, mpt_metadata in self._device_attributes.items():
+        for mpt_name, mpt_metadata in list(self._device_attributes.items()):
             monitoring_pts[mpt_name] = {}
-            for metadata_prop_name, metadata_prop_val in mpt_metadata.items():
+            for metadata_prop_name, metadata_prop_val in list(mpt_metadata.items()):
                 # Unpack the min and max values from the ValueRange dictionary
                 if metadata_prop_name == "ValueRange":
-                    for extremity, extremity_val in metadata_prop_val.items():
+                    for extremity, extremity_val in list(metadata_prop_val.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/sim_sdd_xml_parser.py]: Remove `list`, reasons stated in one of the comments above.

File: tango_simlib/utilities/sim_xmi_parser.py [hunk 11]
```diff
@@ -131,8 +136,8 @@ def parse(self, sim_xmi_file):
         # ensure all unicode attribute values are converted to byte strings
         # as TANGO does not handle unicode
         for child in tree.findall(".//"):
-            for key, value in child.attrib.items():
-                if isinstance(value, unicode):
+            for key, value in list(child.attrib.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/sim_xmi_parser.py]: Ditto

File: tango_simlib/utilities/sim_xmi_parser.py [hunk 12]
```diff
@@ -300,7 +305,7 @@ def extract_attributes_description_data(self, description_data):
         attribute_data["dynamicAttributes"] = description_data.attrib.copy()
 
         attType = attribute_data["dynamicAttributes"]["attType"]
-        if attType in POGO_PYTANGO_ATTR_FORMAT_TYPES_MAP.keys():
+        if attType in list(POGO_PYTANGO_ATTR_FORMAT_TYPES_MAP.keys()):
```
Comment (mmphego) [inline tango_simlib/utilities/sim_xmi_parser.py]: Also here.

File: tango_simlib/utilities/helper_module.py [hunk 13]
```diff
@@ -179,7 +183,7 @@ def json_loads_byteified(json_text):
 
 def _byteify(data, ignore_dicts=False):
     """If this is a unicode string, return its string representation."""
-    if isinstance(data, unicode):
+    if isinstance(data, str):
```
[the remark this answers is not in the document]
Comment (ajoubertza) [reply tango_simlib/utilities/helper_module.py]: Good point. Python 3 `str` is Unicode.
Similar "byteify" code was recently removed from the SKA LMC base classes repo because it now only has to support Python 3.  Looks like PyTango might take care of this in Python 3.  So, you can try only doing the byteify work in Python 2.

File: tango_simlib/utilities/sim_xmi_parser.py [hunk 14]
```diff
@@ -577,8 +582,8 @@ def get_device_attribute_metadata(self):
             for (
                 prop_group,
                 default_attr_props,
-            ) in POGO_USER_DEFAULT_ATTR_PROP_MAP.items():
-                for pogo_prop, user_default_prop in default_attr_props.items():
+            ) in list(POGO_USER_DEFAULT_ATTR_PROP_MAP.items()):
+                for pogo_prop, user_default_prop in list(default_attr_props.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/sim_xmi_parser.py]: Ditto

File: tango_simlib/utilities/sim_xmi_parser.py [hunk 15]
```diff
@@ -633,9 +638,9 @@ def get_device_command_metadata(self):
 
         commands = {}
         # Need to convert the POGO parameter names to the TANGO names
-        for cmd_name, cmd_metadata in temp_commands.items():
+        for cmd_name, cmd_metadata in list(temp_commands.items()):
             commands_metadata = {}
-            for cmd_prop_name, cmd_prop_value in cmd_metadata.items():
+            for cmd_prop_name, cmd_prop_value in list(cmd_metadata.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/sim_xmi_parser.py]: Ditto

File: tango_simlib/utilities/simdd_json_parser.py [hunk 16]
```diff
@@ -169,12 +174,12 @@ def get_device_data_components_dict(self, elements, element_type):
...
         params_template = helper_module.DEFAULT_TANGO_ATTRIBUTE_PARAMETER_TEMPLATE.copy()
         for element_data in elements:
-            for element_info in element_data.values():
+            for element_info in list(element_data.values()):
                 name = element_info["name"]
                 element_params = self._get_reformatted_data(element_info, element_type)
                 if "Attributes" in element_type:
                     device_dict[str(name)] = dict(
-                        params_template.items() + element_params.items()
+                        list(params_template.items()) + list(element_params.items())
```
Comment (mmphego) [inline tango_simlib/utilities/simdd_json_parser.py]: Ditto

File: tango_simlib/utilities/testutils.py [hunk 17]
```diff
@@ -2,11 +2,16 @@
 from __future__ import division
 from __future__ import absolute_import
 
+
```
Comment (mmphego) [inline tango_simlib/utilities/testutils.py]: Move above imports below the copyright info.

File: tango_simlib/utilities/testutils.py [hunk 18]
```diff
@@ -135,7 +140,7 @@ def set_attributes_polling(test_case, device_proxy, device_server, poll_periods)
 
     def restore_polling():
         """Restore initial polling, for use during cleanup / teardown"""
-        for attr, period in initial_polling.items():
+        for attr, period in list(initial_polling.items()):
```
Comment (mmphego) [inline tango_simlib/utilities/testutils.py]: Remove `list`

File: Jenkinsfile [hunk 19]
```diff
@@ -65,26 +64,26 @@ pipeline {
...
+                         sh 'python3.6 -m pip install nose_xunitmp --user'
+                         sh "python3.6 setup.py nosetests --with-xunitmp --with-xcoverage --cover-package=${KATPACKAGE} --with-xunit --xunit-file=nosetests_py36.xml"
                     }
                 }
             }
 
             post {
                 always {
-                    junit 'nosetests.xml'
+                    junit 'nosetests_*.xml'
```
Comment (amakhaba) [inline Jenkinsfile]: @mmphego I removed this because tests failed when generating the reports in the Futurize stage 1 branch please do check it out.
```suggestion
                    junit 'nosetests.xml'
```

File: Jenkinsfile [hunk 20]
```diff
@@ -65,26 +64,26 @@ pipeline {
...
                     cobertura (
-                        coberturaReportFile: 'coverage.xml',
+                        coberturaReportFile: 'coverage_*.xml',
```
Comment (amakhaba) [inline Jenkinsfile]: See above [comment](https://github.com/ska-sa/tango-simlib/pull/108/files#r354695158)
```suggestion
                        coberturaReportFile: 'coverage.xml',
```

**Comment by mmphego** [comment]
@amakhaba please update branch by merging `master` into it and fixing up the conflicts. 

Please also change the target branch from `user/amakhaba/MT-738/Auto-code-format` to `master` as branch `user/amakhaba/MT-738/Auto-code-format` is already in `master` and doesn't exists anymore.

See: [How to modify a pull request on GitHub to change target branch to merge into?](https://stackoverflow.com/questions/24159036/how-to-modify-a-pull-request-on-github-to-change-target-branch-to-merge-into)

File: tango_simlib/sim_test_interface.py [hunk 21]
```diff
@@ -3,28 +3,35 @@
...
 from tango.server import Device, DeviceMeta, attribute, device_property
 from tango_simlib import model
 from tango_simlib.utilities.helper_module import generate_cmd_handler
 
+standard_library.install_aliases()
+
+
 
-class TangoTestDeviceServerBase(Device):
-    __metaclass__ = DeviceMeta
```
Comment (amakhaba) [inline tango_simlib/sim_test_interface.py]: @tockards and @mmphego why was this removed?

Comment (mmphego) [reply tango_simlib/sim_test_interface.py]: It was not removed by an individual but by futurize stage 2.

See: https://python-future.org/compatible_idioms.html?highlight=abc%20metaclass#metaclasses

If indoubt what metaclasses are, check this video out: https://www.youtube.com/watch?v=0NFFk32CJ8s

**Comment by ajoubertza** [comment]
I'm afraid this PR is a bit of mess now - it shows all the changes from earlier PRs as well, including the black autoformatting.  Also a heap of conflicts.  Any ideas?

**Comment by mmphego** [comment]
> I'm afraid this PR is a bit of mess now - it shows all the changes from earlier PRs as well, including the black autoformatting. Also a heap of conflicts. Any ideas?

Seems like there are 2 options (IMHO):

1. Close PR, branch off master, cherry-pick or create new commits and then new PR.
2. Merge master into current branch, change target branch to master and fix conflicts 

But @amakhaba and @tockards can decide.

**Comment by ajoubertza** [comment]
Option 3:  Make a new branch from master and just copy the files from this branch to that one.  Make that a new commit.

Loses the commit history, but clean result and much easier.

**Comment by amakhaba** [comment]
@ajoubertza  and @mmphego option 3 sounds like a better one. Can i finish up with the merge conflicts that i am addressing now and then we can plot a way forward

**Comment by amakhaba** [comment]
@ajoubertza @lanceWilliams @tockards @mmphego I will be closing this PR as it has become to difficult to review please checkout : user/amakhaba/MT-738/Applied-futurize-stage-2 and review for further modifications. For Review please Look at [PR - 110](https://github.com/ska-sa/tango-simlib/pull/110/)

File: tango_simlib/model.py [hunk 22]
```diff
@@ -1,22 +1,29 @@
...
+from builtins import map, object, range
 from functools import partial
 
+from future import standard_library
+standard_library.install_aliases()
+
 from tango import CmdArgType
 from tango_simlib import quantities
 
+
```
Comment (tockards) [inline tango_simlib/model.py]: @amakhaba Please remove redundant spaces.

File: tango_simlib/model.py [hunk 23]
```diff
@@ -1,22 +1,29 @@
...
 import logging
 import sys
 import time
 import weakref
+from builtins import map, object, range
 from functools import partial
 
+from future import standard_library
+standard_library.install_aliases()
+
```
Comment (tockards) [inline tango_simlib/model.py]: @amakhaba take note of adding a `#noqa` for python3
